src/libAtomVM/context.h

```diff
@@ -244,49 +239,57 @@ static inline void context_clean_registers(Context *ctx, int live)
 }
 
 /**
- * @brief Returns available free memory in term units
+ * @brief Returns a context's stack base
  *
- * @details Returns the number of terms that can fit either on the stack or on the heap.
+ * @details Used for stack traces
  * @param ctx a valid context.
- * @returns available free memory that is avail_size_in_bytes / sizeof(term).
+ * @returns the stack base
  */
-static inline unsigned long context_avail_free_memory(const Context *ctx)
+static inline term *context_stack_base(const Context *ctx)
 {
-    return ctx->e - ctx->heap_ptr;
-}
-
-/**
- * @brief Returns context total memory in term units
- *
- * @details Returns the total memory reserved for stack and heap in term units.
- * @param ctx a valid context.
- * @returns total memory that is total_size_in_bytes / sizeof(term).
- */
-static inline unsigned long context_memory_size(const Context *ctx)
-{
-    return ctx->stack_base - ctx->heap_start;
+    // Find which fragment the stack belongs to.
+    if (ctx->e >= ctx->heap.heap_start && ctx->e <= ctx->heap.heap_end) {
+        return ctx->heap.heap_end;
+    }
+    HeapFragment *fragment = ctx->heap.next;
+    while (fragment) {
+        if (ctx->e >= fragment->storage && ctx->e <= fragment->heap_end) {
+            return fragment->heap_end;
+        }
+        fragment = fragment->next;
+    }
+    fprintf(stderr, "Context heap is corrupted, cannot find fragment with stack pointer.\n");
+    AVM_ABORT();
 }
 
 /**
- * @brief Returns context heap size in term units
+ * @brief Returns a context's stack size
  *
+ * @details Return the number of terms currently on the stack. Used for
+ * stack traces.
  * @param ctx a valid context.
- * @returns context heap size in term units
+ * @returns stack size in terms
  */
-static inline unsigned long context_heap_size(const Context *ctx)
+static inline size_t context_stack_size(const Context *ctx)
 {
-    return ctx->heap_ptr - ctx->heap_start;
+    term *stack_base = context_stack_base(ctx);
+    return stack_base - ctx->e;
 }
 
 /**
- * @brief Returns context heap size in term units
+ * @brief Returns available free memory in term units
  *
+ * @details Returns the number of terms that can fit either on the heap.
  * @param ctx a valid context.
- * @returns context heap size in term units
+ * @returns available free memory that is avail_size_in_bytes / sizeof(term).
  */
-static inline unsigned long context_stack_size(const Context *ctx)
+static inline size_t context_avail_free_memory(const Context *ctx)
 {
-    return ctx->stack_base - ctx->e;
+    // Check if stack is on current fragment
```

Comment:
Kind of the same comment -- should this heap-specific code be moved into the memory module?

Reply:
Is it really heap-specific? We could create a function that finds which fragment a pointer belongs to.
But maybe we can keep it in context for now as it only applies to stack pointer?